src/game/client/terraintex.cpp

```diff
@@ -0,0 +1,502 @@
+                        2 * (w + 8));
+                }
+            }
+        }
+    }
+
+    surf->UnlockRect();
+    surf->Release();
+    DX8Wrapper::Handle_DX8_ErrorCode(
+        D3DXFilterTexture(reinterpret_cast<w3dtexture_t>(Peek_Platform_Base_Texture()), nullptr, 0, D3DX_FILTER_BOX));
+
+    if (g_theWriteableGlobalData->m_textureReductionFactor) {
+        reinterpret_cast<w3dtexture_t>(Peek_Platform_Base_Texture())
+            ->SetLOD(g_theWriteableGlobalData->m_textureReductionFactor);
+    }
+
+    return desc.Height;
+#else
+    return 0;
+#endif
+}
+
+bool TerrainTextureClass::Update_Flat(WorldHeightMap *ht_map, int x, int y, int pixels_per_cell, unsigned int cell_width)
+{
+#ifdef BUILD_WITH_D3D8
+    w3dsurface_t surf = nullptr;
+    DX8Wrapper::Handle_DX8_ErrorCode(
+        reinterpret_cast<w3dtexture_t>(Peek_Platform_Base_Texture())->GetSurfaceLevel(0, &surf));
+    D3DSURFACE_DESC desc;
+    memset(&desc, 0, sizeof(desc));
+    DX8Wrapper::Handle_DX8_ErrorCode(surf->GetDesc(&desc));
+
+    captainslog_dbgassert(desc.Width != cell_width * pixels_per_cell, "Bitmap too small.");
+    captainslog_dbgassert(desc.Height != cell_width * pixels_per_cell, "Bitmap too small.");
+
+    if (desc.Width == cell_width * pixels_per_cell) {
+        D3DLOCKED_RECT rect;
+        DX8Wrapper::Handle_DX8_ErrorCode(surf->LockRect(&rect, nullptr, 0));
+
+        if (desc.Format == D3DFMT_A1R5G5B5) {
+            for (int i = 0; i < pixels_per_cell; i++) {
+                for (int j = 0; j < pixels_per_cell; j++) {
+                    unsigned char *data = ht_map->Get_Pointer_To_Tile_Data(i + x, j + y, cell_width);
+
+                    if (data) {
+                        for (int k = cell_width - 1; k >= 0; k--) {
+                            unsigned char *data3 = (unsigned char *)rect.pBits;
+                            short *data2 = (short *)&data3[2 * desc.Width * (k + (pixels_per_cell - j - 1) * cell_width)
+                                + 2 * cell_width * i];
+
+                            for (int l = 0; l < (int)cell_width; l++) {
+                                *data2 =
+                                    ((int)*data >> 3) + ((int)data[2] >> 3 << 0xA) + 0x20 * ((int)data[1] >> 3) + 0x8000;
+                                data2++;
+                                data += 4;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        surf->UnlockRect();
+        surf->Release();
+        DX8Wrapper::Handle_DX8_ErrorCode(
+            D3DXFilterTexture(reinterpret_cast<w3dtexture_t>(Peek_Platform_Base_Texture()), nullptr, 0, D3DX_FILTER_BOX));
+        return desc.Height != 0;
+    }
+#endif
+    return false;
+}
+
+void TerrainTextureClass::Set_LOD(int lod)
+{
+#ifdef BUILD_WITH_D3D8
+    if (Peek_Platform_Base_Texture()) {
+        reinterpret_cast<w3dtexture_t>(Peek_Platform_Base_Texture())->SetLOD(lod);

```

Comment:
We probably need a better solution for this so we don't have to do this every time
@OmniBlade ideas?